feat(payments): logic for sending emails subscription invoices & cancellation

lmorchard opened the pull request:
- send initial and subsequent invoice emails from new
  invoice.payment_succeeded webhook event handler

- send subscription cancellation emails from
  customer.subscription.deleted webhook event handler

- split out updateCustomer from sending SQS subscription update message

- allow updateCustomerAndSendStatus to reuse uid & email from a previous
  API call, if available

FXA-1417

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -856,6 +856,99 @@ class StripeHelper {
+   * @param {string | object} resource
+   * @param {string} resourceType
+   *
+   * @returns {Promise<object>}
```
A generic can be used here and in the param above instead of `object`, which should allow a call-in to specify the expected return type and ensure it has it. Since the Stripe lib is already going to be passing this in as `string | StripeType` that should 'just work' and get the right type out.

bbangert replied:
Also, clever with this function! I wish we had this in earlier!

lmorchard replied:
Hmm, the `@template` works here. But, still need some casting it seems. The Stripe functions don't simply return Product & Customer. Instead I get this type:
```
[
Stripe.Product | Stripe.DeletedProduct,
Stripe.Customer | Stripe.DeletedCustomer,
Stripe.Charge
]
```
And doing an `if (!product.deleted && !customer.deleted)` guard doesn't seem enough of a hint to the type inference.

bbangert replied:
Indeed, I think this would only work correctly in TypeScript, the additional features needed to make it work correctly can't be fully embedded in JSDoc.

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -856,6 +856,99 @@ class StripeHelper {
+          product,
+          customer,
+          charge,
+        ] = /** @type {[ Product, Customer, Charge ]} */ (await Promise.all([
```
If the generic above works 🤞 , then the type assert here can go.

lmorchard commented:
Looks like adding two separate `deleted` guard conditions makes the type inference more happy: 
```
        // We shouldn't end up with either a deleted customer or product,
        // but guarding here might log more informative errors just in case
        if (customer.deleted === true) {
          throw error.unknownCustomer(invoice.customer);
        }
        if (product.deleted === true) {
          throw error.unknownSubscriptionPlan(plan.id);
        }
```

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -856,6 +856,106 @@ class StripeHelper {
+        // Dig up & expand objects in the invoice that usually come as just IDs
+        const { plan } = invoice.lines.data[0];
+        const [product, customer, charge] = await Promise.all([
+          this.expandResource(plan.product, 'products'),
```
We can't locate this in the product cache?

lmorchard replied:
Can probably use the allProducts cache here

lmorchard replied:
Reworked things to use AbbrevProduct, so we should hit cache nearly all the time here.

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -856,6 +856,106 @@ class StripeHelper {
+        const { plan } = invoice.lines.data[0];
+        const [product, customer, charge] = await Promise.all([
+          this.expandResource(plan.product, 'products'),
+          this.expandResource(invoice.customer, 'customers'),
```
Would this be in the customer cache?

lmorchard replied:
No - our customer cache is keyed on FxA uid & email. We only have a customer ID here.

lmorchard replied:
We might want to add a follow-up issue to start caching customers on Stripe ID, which would let us use cache here. But, I also kind of think these emails would be a decent spot to have fresh data?

bbangert replied:
The Redis cache should be up to date, if its not then we have a bug.

bbangert commented on packages/fxa-auth-server/lib/routes/subscriptions.js:
```diff
@@ -620,6 +626,23 @@ class DirectStripeRoutes {
+    const sub = await this.stripeHelper.expandResource(
+      invoice.subscription,
+      'subscriptions'
+    );
```
How come this doesn't fetch the customer (too bad we aren't dual keying our cache to stripe customer string) here? Also, it appears that when a payment is succeeded, we're about to send out a subscription state change telling our RP's to de-activate the subscription? (the 'false' in the call below). We should check the existing status of the customer before sending out subscription state changes, since someone that always pays on time isn't having their subscription state changed.

I believe there is a function we have that can refresh a customer given their uid/email, and send out a SQS message if something changed. That feels like a good fit here. (Assuming this function I remember exists)

lmorchard replied:
>How come this doesn't fetch the customer

Customer isn't needed here - `updateCustomerAndSendStatus` wants a subscription and fetches customer from cache itself.

> Also, it appears that when a payment is succeeded, we're about to send out a subscription state change telling our RP's to de-activate the subscription?

Switching this to `true`

> I believe there is a function we have that can refresh a customer given their uid/email, and send out a SQS message if something changed. That feels like a good fit here.

Isn't that what `updateCustomerAndSendStatus` is doing with `refreshCachedCustomer` and `profile.deleteCache`?

bbangert replied:
> Customer isn't needed here - updateCustomerAndSendStatus wants a subscription and fetches customer itself from subscription info.

Right, I mean, we could possibly avoid 1 stripe call by having a function that can use a uid/email instead to fetch from our Redis customer cache and match the subscription id. It would need another function instead of `updateCustomerAndSendStatus`, so I guess it depends on how important it is to reduce the stripe API calls here.

> Isn't that what updateCustomerAndSendStatus is doing with refreshCachedCustomer and profile.deleteCache?

Nope, `updateCustomerAndSendStatus` will always send a SQS event. The `handleSubscriptionUpdatedEvent` function is using the prior attributes to determine if a subscription state change event is needed.

Upon further reflection walking my 🐕 , it occurs to me that we should not be sending SQS events based on invoice payment results (I doubt we need to update our cache either). When an invoice fails, that doesn't actually mean they immediately lose access to the product. An invoice can fail for many reasons, and may be retried multiple times before success. When it can no longer be invoice'd, the subscription is cancelled/deleted, and our webhook handler for that event will propagate the subscription state change. (flagging @jackiemunroe to double-check me on this, but I believe this is correct).

lmorchard replied:
> Right, I mean, we could possibly avoid 1 stripe call by having a function that can use a uid/email instead to fetch from our Redis customer cache and match the subscription id.

But in this invoice payment hook, we don't have access to a full customer record and so don't have an FxA UID at this point.

bbangert replied:
Oh, I meant that if we fetch the customer from Stripe, we'll then have the uid/email. That will be 1 call (if they're then in Redis), instead of the 2 calls this has right now.

jackiemunroe replied:
@bbangert you are correct. Once Stripe has deemed that an invoice is not collectable, there will be a state change on the invoice and the subscription, both of which will trigger webhooks.

jackiemunroe replied:
But, I believe that the Subscription would move to an "incomplete_expired" state

bbangert replied:
It looks like "incomplete_expired" is only if its the very first payment. Recurring payments will go to "past_due" or be cancelled depending on settings (ours indicate it goes to cancelled).

lmorchard commented:
Pushed an update that switches to cached AbbrevProduct details with a last-ditch fallback to a Stripe fetch if it's missing (should be unlikely). Also removed the `updateCustomerAndSendStatus` call from invoice emails altogether - that state change should get picked up in the existing subscription state change handlers.

lmorchard commented:
> it occurs to me that we should not be sending SQS events based on invoice payment results (I doubt we need to update our cache either)

@bbangert I removed the cache update - but per triage convo, should I re-add?

bbangert commented:
@lmorchard  Yes please, the refreshing the cache here would remedy the issue for monthly subs.

lmorchard commented:
Okay: Pushed a some new changes to update the customer from the invoice payment success hook, but without sending an SQS message. Also cut back on a few API calls by re-using the customer fetch needed for sending the subscription cancellation email

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.js:
```diff
@@ -856,6 +867,137 @@ class StripeHelper {
+          ...plan.metadata,
+        };
+        uid = customer.metadata.userid;
+        ({ product_id: productId, product_name: productName } = abbrevProduct);
```
Why did these get wrapped in a set of parens? The destructuring doesn't work otherwise?

lmorchard replied:
Yes, destructuring into local vars requires parens like this.

chenba replied:
`const { product_id: productId, product_name: productName } = abbrevProduct;`?

lmorchard replied:
No, these are being assigned to variables declared earlier with `let` https://github.com/mozilla/fxa/pull/4704/files#diff-26644f246b5f22f6205561269d539ce7R885
